subprojects/junit-jupiter/src/test/java/org/mockitousage/GenericTypeMockTest.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2023 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+
+package org.mockitousage;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class GenericTypeMockTest {
+
+    public static class UnderTest {
+        List<String> stringList;
+        List<Integer> intList;
+        Set<?> anySet;
+        Map<Integer, ? extends Collection<String>> intStringCollectionMap;
+    }
+
+    @Mock
+    private List<String> stringProviderMock;
+
+    @Mock
+    private List<Integer> intProviderMock;
+
+    @Mock
+    private TreeSet<String> treeSetMock = Mockito.mock(TreeSet.class);;
+
+    @Mock
+    private Map<Integer, List<String>> intStringListMapMock;
+
+    @InjectMocks
+    private UnderTest underTest;
+
+    /**
+     * Verify that InjectMocks will correctly match fields with same generic type but different type parameters,
+     * without using the same field name.
+     */
+    @Test
+    public void testInjectMock() {
+        // this used to fail without any error message hinting at the problem, as soon as a class under test has
+        // a second field of the same generic type, but with different type parameter. The programmer then
+        // had to know that mock field names have to match field names in the class under test.
+        assertNotNull(underTest.stringList);
```

Comment:
Please define a separate test case for each of these, preferably with an inner class with each Mock, so that it is clear which mock is used in which class

Reply:
Done